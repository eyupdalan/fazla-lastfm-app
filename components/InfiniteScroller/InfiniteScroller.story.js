import InfiniteScroller from "./index";
import {QueryClient, QueryClientProvider} from "react-query";

const infiniteScrollerStory = {
    title: 'Components/InfiniteScroller',
    component: InfiniteScroller,
};

const mockApiCallback = async () => {
    const arr = [];
    for (let i = 0; i < 50; i++) {
        arr.push({x: Math.random()});
    }
    return arr;
};

const queryClient = new QueryClient();

export const InfiniteScrollerStory = () => (
    <QueryClientProvider client={queryClient}>
        <div style={{width: 300, height: 400, border: "1px solid black", overflow: "scroll" }}>
            <InfiniteScroller
                queryKey={"test"}
                apiCallback={mockApiCallback}
                childrenRenderer={(data) => {
                    if (!data) {
                        return null;
                    }

                    return data.pages.map(page=>{
                        return page.map(item=>{
                            return <div key={item.x} style={{borderBottom: "1px solid gray", padding: 5}}>{item.x}</div>
                        })
                    })
                }}
                lastPageController={() => false}
            />
        </div>
    </QueryClientProvider>
);

export default infiniteScrollerStory;