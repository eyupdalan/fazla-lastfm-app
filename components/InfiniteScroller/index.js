import {useEffect, useRef, useState} from "react";
import PropTypes from "prop-types";
import {useInfiniteQuery} from "react-query";

export default function InfiniteScroller({queryKey, apiCallback, childrenRenderer, lastPageController}) {
    const [isLastPage, setIsLastPage] = useState(false);
    const [lastElement, setLastElement] = useState(null);
    const observer = useRef(null);

    const getNextPageParam = (lastPage) => {
        return lastPage.page + 1
    }
    const {
        isLoading,
        isError,
        error,
        data,
        fetchNextPage,
        isFetching,
        isFetchingNextPage
    } = useInfiniteQuery([queryKey], apiCallback, {getNextPageParam});

    useEffect(()=>{
        setIsLastPage(lastPageController(data))
<<<<<<< HEAD
    },[data])
=======
    },[data, lastPageController])
>>>>>>> 4615feb (eyup adds tests and stories)

    useEffect(() => {
        const currentElement = lastElement;
        const currentObserver = observer.current;

        if (currentElement) {
            currentObserver.observe(currentElement);
        }

        return () => {
            if (currentElement) {
                currentObserver.unobserve(currentElement);
            }
        };
    }, [lastElement]);


    useEffect(() => {
        const generateObserver = (entries) => {
<<<<<<< HEAD
            debugger;
=======
>>>>>>> 4615feb (eyup adds tests and stories)
            if (isLastPage) {
                return;
            }
            const first = entries[0];
            if (first.isIntersecting) {
                fetchNextPage();
            }
        }

        observer.current = new IntersectionObserver(generateObserver);
    }, [isLastPage, fetchNextPage]);

    return (
        <>
            <div>{isLoading ? 'Loading...' : null}</div>
            {childrenRenderer(data)}
            <div ref={setLastElement}/>
            <div>{isError ? error.message : null}</div>
            <div>{isFetching && !isFetchingNextPage ? 'Fetching...' : null}</div>
        </>
    )
}

InfiniteScroller.propTypes = {
    queryKey: PropTypes.string.isRequired,
    apiCallback: PropTypes.func.isRequired,
    childrenRenderer: PropTypes.func.isRequired,
    lastPageController: PropTypes.func.isRequired
}

InfiniteScroller.defaultProps = {
    queryKey: '',
    apiCallback: () => {
    },
    childrenRenderer: () => {
    },
    lastPageController: () => {
    }
}