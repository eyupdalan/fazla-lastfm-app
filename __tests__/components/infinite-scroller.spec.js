import {render} from '@testing-library/react'
import InfiniteScroller from '../../components/InfiniteScroller';
import {QueryClient, QueryClientProvider,} from 'react-query'

const mockApiCallback = async () => {
    return [{item: Math.random()}];
};
const queryClient = new QueryClient();
const wrapper = ({children}) => (
    <QueryClientProvider client={queryClient}>
        {children}
    </QueryClientProvider>
);

/**
 * https://stackoverflow.com/a/58651649/6593949
 *
 * Utility function that mocks the `IntersectionObserver` API. Necessary for components that rely
 * on it, otherwise the tests will crash. Recommended to execute inside `beforeEach`.
 * @param intersectionObserverMock - Parameter that is sent to the `Object.defineProperty`
 * overwrite method. `jest.fn()` mock functions can be passed here if the goal is to not only
 * mock the intersection observer, but its methods.
 */
export function setupIntersectionObserverMock({
                                                  root = null,
                                                  rootMargin = '',
                                                  thresholds = [],
                                                  disconnect = () => null,
                                                  observe = () => null,
                                                  takeRecords = () => [],
                                                  unobserve = () => null,
                                              } = {}) {
    class MockIntersectionObserver {
        constructor() {
            this.root = root;
            this.rootMargin = rootMargin;
            this.thresholds = thresholds;
            this.disconnect = disconnect;
            this.observe = observe;
            this.takeRecords = takeRecords;
            this.unobserve = unobserve;
        }
    }

    Object.defineProperty(window, 'IntersectionObserver', {
        writable: true,
        configurable: true,
        value: MockIntersectionObserver
    });

    Object.defineProperty(global, 'IntersectionObserver', {
        writable: true,
        configurable: true,
        value: MockIntersectionObserver
    });
}

describe('Testing Infinite Scroller component', function () {
    beforeEach(()=>{
        setupIntersectionObserverMock();
    })

    it('renders Artist component', () => {
        const {container} = render(
            <InfiniteScroller
                queryKey={"test"}
                apiCallback={mockApiCallback}
                lastPageController={() => false}
                childrenRenderer={(data) => data ? data.pages.map(item => (<div>item</div>)) : null}
            />,
            {wrapper}
        )
        expect(container).toMatchSnapshot()
    })
});