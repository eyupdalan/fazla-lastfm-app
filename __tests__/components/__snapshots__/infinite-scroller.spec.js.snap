// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Infinite Scroller component renders Artist component 1`] = `
<div>
  <div>
    Loading...
  </div>
  <div />
  <div />
  <div>
    Fetching...
  </div>
</div>
`;
